### DISCUSSION TITLE
Combining CTX with fabric.js

### DISCUSSION THREAD

**Discussion by jeroenrosendaal** [fabricjs/fabric.js#10158] [General]
Hi all,

can we combine drawing with fabric.js and CTX on the same canvas?

When we try this the graphics drawn with CTX are removed when adding a fabric.js object to the canvas.

```js
const canvas = document.getElementById('canvas');
canvas.width = document.body.clientWidth;
canvas.height = document.body.clientHeight;

const fabricCanvas = new fabric.Canvas('canvas');

const ctx = canvas.getContext('2d');
ctx.fillStyle = "green";
ctx.fillRect(10, 10, 150, 150);

const rect = new fabric.Rect({
  left: 100,
  top: 100,
  fill: '#000',
  width: 20,
  height: 40,
});


// when commented the green rectangle will be drawn,
// otherwise the black rectangle.
fabricCanvas.add(rect);
```

**Comment by asturur** [comment]
FabricJS offer an after-render event and you can draw your ctx code if you really have to.
You shouldn't in my opinion and try to wrap your ctx code in a class that goes along with the fabricJS rendering.
With the snippet you provided i would tell you, just add another rect, but also you could render that in the after:render or before:render events.

**Reply by jeroenrosendaal** [reply]
I'm trying to use another library along side fabric.js called wordcloud2.js so we don't have to implement our own logic for the wordcloud. Would you suggest to do this wrap this in a class? Would that be a subclass of the fabric.Object class?

**Reply by asturur** [reply]
What are you doin with wordcloud? And what functionality you expect?

This is what i see on wordcloud website
<img width="1189" alt="image" src="https://github.com/user-attachments/assets/49688b43-eca8-42ca-8807-59408cd0edab">

This for me could be a way to lay text on a canvas OR have a nice layer for a design or whatever.
In the first case i would want to change all those single words in singe fabricJS text so they can still be moved/edited individually. in the other case i would probably just make a class that is similar to Image for it.

It mostly depends what you need.
Do you need to change the state of wordcloud and have the fabricCanvas update quickly? I would go with a custom class extending Object and looking in to how Rect is built to have a guideline.
Do you need to have it static and change only sometimes the colors or basic scaling interactivity? i would probably treat it as an image and regenerate it with a different color/resolution when necessary.

In general if you need to have it mixed with the object stack you need it to be a custom class.

**Reply by jeroenrosendaal** [reply]
Your first case I have tried but due to the padding on the Text objects the positions is a little off so the text will overlap in certain cases. I did not find a solution to remove this 'padding'. I'm able to retrieve the coordinates of the words. Do you know if there is a way to remove the 'padding'? I don't think is called padding but more the line height?

**Reply by asturur** [reply]
There is no functionality to remove the small padding around text. You can reduce lineHeight to 1 but this won't reduce the left/right padding you may have.

**Reply by jeroenrosendaal** [reply]
Then I will try the subclassing approach first. Thanks for your time.

**Reply by asturur** [reply]
The subclassing approach is easier than you think.

**Reply by jeroenrosendaal** [reply]
I used it before but my concern is that the wordcloud will be rebuild every time the canvas renders. Is that true?

**Reply by asturur** [reply]
That is true only if you build in that way.
The best thing to do is to render the world cloud only when the world cloud property changes.
That is your class it has to do what you want.
So for example on init can create a canvas cloud with the size you want, and keep it on an internal canvas as buffer, on render it will display that canvas like if it was a simple jpeg. Then when you decide to change the content of the cloud you will regenerate that and refresh your buffer canvas.

If you do this manually you have control on scaling and zooming and how often do you want it to change.

**Reply by jeroenrosendaal** [reply]
Thanks. On the mousedown event I would like to pass down the current canvas so wordcloud2.js can do it's own thing, like returning the word that is clicked on. I've created a jsbin. Would you mind taking a look at it? 

This peace of code is copied from worcloud.js. If I create a custom `PointerEvent` and pass down the `currentTarget` the `DOMRect` will be empty (everything stays at 0)
```js
this.tmpCanvas.addEventListener('click', (e) => {
    const target = e.currentTarget;
    var rect = target.getBoundingClientRect();

    //EXPECTED RESULT: A filled in DOMrect
    console.log(rect);
});
```
Do you know why it says the tempCanvas has no dimensions?

[https://jsbin.com/zihanaj/edit?html,js,console,output](https://jsbin.com/zihanaj/edit?html,js,console,output)

**Reply by jeroenrosendaal** [reply]
@asturur Do you have an idea about how to pass down the mousedown event (or other events) to the underlying canvas so that the other library can do it's own thing when the event is fired?

**Reply by asturur** [reply]
If the library doesn't expose a function to do so you have to forward the event yourself.

Is not about CTX anymore at this point, is it?

You have to do something like:

```
this.tmpCanvas.addEventListener('click', (e) => {
   const mouseEvent = new MouseEvent(e.type, e);
   yourCanvasHTMLElement.dispatch(mouseEvent);
}
```

Not sure of the exact code to write

**Reply by jeroenrosendaal** [reply]
Thanks for your reply. I've already implemented this but the DOMRect was still not filled in. The problem is that the canvas needed to be added to the DOM before it will return coordinates. I modified the src code of wordcloud2.js so it will accept an object with and x, y, w, h, property. This will do the trick for now. Thanks for your time.

**Reply by asturur** [reply]
Oh ok because their internal functions requires a dom rect from the event, as fabricjs does too.
You could open a PR so that it exposes a function that given a X and Y will tell you the correct word,
